made ltsp server flexible

d78ui98 opened the pull request:
- [x]  Added `settings.sh`
- [x]  Made changes in `install.sh` to load mode of operation from `settings.sh` 
- [x]  small bug fix to to allow standalone mode of operation
- [x] changes in `Vagrantfile` to get network address 
- [x] update `README.md`

fixes #28

d78ui98 committed:
making flexible ltsp server

added setting.sh It is loaded by install.sh that sets the mode of operation on ltsp server

dashohoxha commented on install.sh:
```diff
@@ -45,3 +45,31 @@ apt-get --yes install ltsp-manager
+sed -i "s/dhcp-range=192.168.67.20,192.168.67.250,8h/dhcp-range=192.168.1.20,192.168.1.250,8h/g" /etc/dnsmasq.d/ltsp-server-dnsmasq.conf
+service dnsmasq restart	
+
+#Reading setting.sh file
```
Try `source settings.sh` instead. This is like including that file.

dashohoxha commented on install.sh:
```diff
@@ -45,3 +45,31 @@ apt-get --yes install ltsp-manager
+response=($(echo "$var" | tr '=' '\n'))
+
+#Setting mode of operation of ltsp server
+if [ "${response[1]}" == "Yes" ] || [ "${response[1]}" == "yes" ]; then
```
Use the variable `$STANDALONE`.

dashohoxha commented:
Does it work if you set `NETWORK=192.168.100.0/24` on `settings.sh`?

d78ui98 committed:
updated Vagrantfile install.sh

use of source implemented in install.sh. settings.sh file is loaded in Vagrantfile. dhcp-range takes value from settings.sh. updated README.md

dashohoxha commented on README.md:
```diff
@@ -1,27 +1,33 @@
+User can put yes or no depending upon the mode of operation. Also you can put network address and LAN IP. Settings will be loaded by `Vagrantfile` and `install.sh` Then you can do a `vagrant up` and Vagrant will automatically install LTSP with the help of provisioning script. 
+
+*Note* - If you have 2 network interfaces then after doing `vagrant up` it will ask which interface you want the vagrant box to bridge with. Selected the one which has internet connection with it. 
+
```
You should try to automate this with scripts, so that it does not need to ask the user which interface to select. I think that @rr83019  had some ideas about this. But you can also search in the internet for some possible solutions.

dashohoxha commented on Vagrantfile:
```diff
@@ -1,11 +1,23 @@
+
+obj = Config.new()
+obj.readfunction
+
```
Why do you make things so complicated? You can use `load settings.sh`, which includes that file in the ruby script (similar to `source settings.sh` which includes it in a bash script). I don't know ruby well, but I searched and I found this solution.

Then, in `settings.sh`, you can use `NETWORK="192.168.10"`, instead of `NETWORK=192.168.10`, so that it is valid both in ruby and in bash. Also be aware that you should not use spaces around `=`, otherwise it will not be valid in bash.

dashohoxha commented on install.sh:
```diff
@@ -15,61 +13,52 @@ apt-get --yes install epoptes epoptes-client
+#source setting.sh
+source /vagrant/settings.sh
+NET=$(echo $NETWORK | cut -d'.' -f 3)
+
```
Make the format of `NETWORK` in `settings.sh` what you need it to be, so that you don't have to process it like this. For example make it `NETWORK="192.168.10"` (with double quotes).

dashohoxha commented on install.sh:
```diff
@@ -15,61 +13,52 @@ apt-get --yes install epoptes epoptes-client
+NET=$(echo $NETWORK | cut -d'.' -f 3)
+
+#setting dhcp-range
+sed -i 's/dhcp-range=192.168.67.20,192.168.67.250,8h/dhcp-range=192.168.'"$NET"'.20,192.168.'"$NET"'.250,8h/g' /etc/dnsmasq.d/ltsp-server-dnsmasq.conf
```
This command can be more readable like this:
```
sed -i /etc/dnsmasq.d/ltsp-server-dnsmasq.conf \
    -e "/^dhcp-range=/ c dhcp-range=${NETWORK}.20,${NETWORK}.250,8h"
```

Writing beautiful and readable code is an art that you should try actively to learn.

d78ui98 replied:
This changes multiple lines. will be using 
```
sed -i /etc/dnsmasq.d/ltsp-server-dnsmasq.conf -e "/8h$/ c dhcp-range=${NETWORK}.20,${NETWORK}.250,8h"
```
instead to uniquely identify this particular line. I will learn how to write code readable

dashohoxha commented on install.sh:
```diff
@@ -15,61 +13,52 @@ apt-get --yes install epoptes epoptes-client
 #Setting mode of operation of ltsp server
-if [ "${response[1]}" == "Yes" ] || [ "${response[1]}" == "yes" ]; then
+if [ "$STANDALONE" == "Yes" ] || [ "$STANDALONE" == "yes" ]; then
 	echo "LTSP server will be in standalone mode of operation"	
```
If you convert the value of `$STANDALONE` to lowercase (or uppercase), you can make only one test. For example:
```
if [[ ${STANDALONE,} == "yes" ]]; then
```
There are a lot of tricks that you can learn about bash scripting. It seems that you have not studied very carefully the book that I told you, or you have missed some important parts of it.

d78ui98 replied:
>It seems that you have not studied very carefully the book that I told you, or you have missed some important parts of it.

sorry about that.
can I use `if [[ ${STANDALONE,,} == "yes" ]]; then`   instead
It can handle case of `YES` along with `Yes`.

dashohoxha commented on install.sh:
```diff
@@ -15,61 +13,52 @@ apt-get --yes install epoptes epoptes-client
-	sed -i "16 s/dhcp-range=192.168.1.0,proxy/#dhcp-range=192.168.1.0,proxy/g" /etc/dnsmasq.d/ltsp-server-dnsmasq.conf
-	service dnsmasq restart
-elif [ "${response[1]}" == "No" ] || [ "${response[1]}" == "no" ]; then
+	sed -i '16 s/dhcp-range=192.168.1.0,proxy/#dhcp-range=192.168.1.0,proxy/g' /etc/dnsmasq.d/ltsp-server-dnsmasq.conf
```
Making changes on a fixed line number is fragile. The default dnsmasq configuration may change for some reason (for example on the next release). The you script will break. Try to make this more robust.

d78ui98 committed:
updated install.sh Vagrantfile

use of load implemented in Vagrantfile. Also improved readibility of the files.

d78ui98 commented:
Thanks @dashohoxha. Code now looks readable, beautiful and far more efficient :)

d78ui98 committed:
automatic bride interface selection

made some changes in Vagrantfile now it does not ask for which interface to bridge to. Made changes in readme.md file accordingly

dashohoxha commented on Vagrantfile:
```diff
@@ -1,11 +1,14 @@
 # vi: set ft=ruby :
+load "settings.sh"
+
+$IP = NETWORK + ".18".to_s
```
- Do you need `.to_s` here?
 - Do you need `$`?
 - Why do you hard-code the number `.18`?
 - Why don't you use `LAN_IP`?

dashohoxha commented on install.sh:
```diff
@@ -15,33 +13,50 @@ apt-get --yes install epoptes epoptes-client
+source /vagrant/settings.sh
+
+#setting dhcp-range
+sed -i /etc/dnsmasq.d/ltsp-server-dnsmasq.conf -e "/8h$/ c dhcp-range=${NETWORK}.20,${NETWORK}.250,8h"
```
- Careful with the `$` here: `/8h$/`, you need to escape it.
 - Maybe you can be a bit more specific, like this: `-e "/^dhcp-range=.*,8h\$/ c .....`
 - Split long commands into several lines by using `\`

dashohoxha commented on install.sh:
```diff
@@ -15,33 +13,50 @@ apt-get --yes install epoptes epoptes-client
+	echo "LTSP server will be in standalone mode of operation"	
+	echo "LTSP server will provide DHCP services.."	
+	sed -i /etc/dnsmasq.d/ltsp-server-dnsmasq.conf -e "/192.168.1.0,proxy$/ c #dhcp-range=${NETWORK}.0,proxy"
+	sed -i /etc/dnsmasq.d/ltsp-server-dnsmasq.conf -e "/10.0.2.0,proxy$/ c #dhcp-range=10.0.2.0,proxy"
```
Try to merge these 2 `sed` commands into one.

dashohoxha commented on install.sh:
```diff
@@ -15,33 +13,50 @@ apt-get --yes install epoptes epoptes-client
+	echo "LTSP server will provide DHCP services.."	
+	sed -i /etc/dnsmasq.d/ltsp-server-dnsmasq.conf -e "/192.168.1.0,proxy$/ c #dhcp-range=${NETWORK}.0,proxy"
+	sed -i /etc/dnsmasq.d/ltsp-server-dnsmasq.conf -e "/10.0.2.0,proxy$/ c #dhcp-range=10.0.2.0,proxy"
+elif [[ ${STANDALONE,,} == "no" ]]; then
```
Make this just `else`. So the default will be non-standalone for any value different from `yes` or `Yes` or `YES`.

dashohoxha commented on settings.sh:
```diff
@@ -0,0 +1,3 @@
+STANDALONE="yes"
+NETWORK="192.168.1"
+LAN_IP="192.168.1.1/24"
```
Do you use `LAN_IP` anywhere?

dashohoxha commented:
Please also explain what steps you follow for testing this in order to make sure that everything works as expected. So that the other people can test and verify this.

d78ui98 committed:
updated install.sh and Vagrantfile

used LAN_ip. Removed hardcoded number. merged 2 sed into one. removed elif condition

dashohoxha commented:
> Testing steps -

No client is mentioned here. You cannot test a ltsp-server without running a client. Where is your client?

Actually I would recommend that you create a document that describes (to the new users) the standalone operation mode, including also some network diagrams. Then describe also the dependent (non-standalone) operation mode, again with some diagrams.
And then a document that describes how to test each of these modes.

You can use the format markdown for these documents. It can be something similar to this: https://github.com/EasyGnuPG/ggpg/blob/describing-application/docs/description.md

For drawing network diagrams you can use Dia or some other tool that you are familiar with. I have used Dia many years ago and I have been happy with it. For example have a look at the network diagrams here:
 - https://www.researchgate.net/publication/317181126_Building_a_Gateway_Server

d78ui98 commented:
Sounds good. Should I create a separate branch for this?

dashohoxha commented:
> Should I create a separate branch for this task?

You can create a separate branch if you wish, or you can try using wiki:
https://github.com/docker-scripts/dev--LTSP/wiki

In any case don't forget to come back and finish the existing branches:
https://github.com/docker-scripts/dev--LTSP/pulls

d78ui98 commented:
okay. I think using wiki would be a good idea for this

akash0x53 commented on Vagrantfile:
```diff
@@ -2,13 +2,11 @@
   config.vm.box = "ubuntu/bionic64"
 	
-  config.vm.network "public_network", ip: $IP, netmask: "255.255.255.0", bridge: "eno1"
+  config.vm.network "public_network", ip: LAN_IP, netmask: "255.255.255.0", bridge: "eno1"
```
vagrant will ask user to select interface if `eno1` not found on host machine. So this also need to put in settings.sh, also `netmask`?

dashohoxha replied:
Let's say that netmask is always `255.255.255.0` (only in very rare cases you may need to make it different).

About `eno1`, I think that there should be some way to discover it automatically.

Also, if there are several network interfaces, the script should be able to autodetect which one is connected to the internet (for example the wifi interface) and use it as the WAN interface of the ltsp box.

But if something cannot possibly be discovered automatically, then yes, we should put it on `settings.sh` instead of hardcoding it on the scripts.

d78ui98 replied:
good point. I think the best solution could be to fetch interface name from the host machine. On first thoughts this would be helpful `VBoxManage list bridgedifs | grep -m1 Name | awk '{print $2}'` But I am sure there are better ways doing same. Will search for that.
Also should I write seperate line for `netmask` in `settings.sh`?

dashohoxha replied:
> Also should I write seperate line for netmask in settings.sh?

You don't need to, since it is always fixed to `255.255.255.0`.

dashohoxha replied:
> On first thoughts this would be helpful VBoxManage list bridgedifs

Try also the commands `ip address` and `ip link`.

d78ui98 committed:
fetch interface name

added a line in Vagrantfile that scrapes the name of interface connected to the internet

d78ui98 commented:
I implemented same with `ip address` and `ip link` command too. But, these commands don't actually tell whether interface is connected to internet or not. So I thought best option is use `route`

code is tested multiple times. Please let me if we need to make other improvements or if we can merge this

d78ui98 committed:
using ip route

using ip route instead of route and grep intead of awk

dashohoxha commented:
We can merge it if you want. But it is not tested unless you test it with a client: https://github.com/docker-scripts/dev--LTSP/pull/31#issuecomment-393690718

Just because the virtual box boots up and the scripts run without errors, it doesn't mean that it is configured and works correctly. When you say "made ltsp server flexible" it means that it should work both for standalone and non-standalone case. You have not tested properly none of these cases.

Anyway let us merge it now and you can continue to fix and improve it later.
